http/src/main/java/io/micronaut/http/filter/FilterRunner.java

```diff
@@ -941,168 +748,256 @@ private DelayedFilterReturnHandler(boolean isResponseFilter, FilterReturnHandler
+                    } catch (Throwable e) {
+                        return ExecutionFlow.error(e);
                     }
                 });
-                return false;
             }
         }
     }
 
-    /**
-     * Some filters are called for the request <i>and</i> the response (e.g. legacy, or "around"
-     * filters), and need to carry state between the two. To achieve this, they are called once for
-     * the request, where they register a SuspensionPoint. The SuspensionPoint is then completed
-     * when the filter is reached for response processing. The filter has a listener on the
-     * SuspensionPoint and handles further processing.<br>
-     * There is also a special suspension point with filter index -1 which manages the flow
-     * returned by {@link #run(HttpRequest)}.
-     */
-    private static class SuspensionPoint extends CompletableFuture<HttpResponse<?>> {
-        /**
-         * The index of the filter this suspension point is associated with.
-         */
-        final int filterIndex;
-        /**
-         * The next suspension point after this one. The invariant is that
-         * {@code this.filterIndex > next.filterIndex}.
-         */
-        @Nullable
-        final SuspensionPoint next;
-
-        SuspensionPoint(int filterIndex, @Nullable SuspensionPoint next) {
-            this.filterIndex = filterIndex;
-            this.next = next;
-        }
-    }
-
     /**
      * This class implements the "continuation" request filter pattern. It is used by filters that
      * accept a {@link FilterContinuation}, but also by legacy {@link HttpFilter}s.<br>
-     * Continuations give the user the choice when to proceed with filter execution. This adds some
-     * difficulty for concurrency: we must ensure {@link #workRequest()} (or
-     * {@link #workResponse()} in case of errors) is called exactly once. For this purpose, this
-     * class has the {@link #downstreamGuard} flag. It is set to {@link true} when the filter
-     * calls {@link #proceed}, or when the filter throws an exception. Only the first of these
-     * events actually continues working the {@link FilterRunner}, any events that follow are
-     * discarded (logged or throw an exception).
+     * Continuations give the user the choice when to proceed with filter execution.
+     * After the proceed is triggered the filter is essentially suspended and the next filter in the chain should be executed.
      *
      * @param <R> Return value of the continuation
      */
-    private abstract class FilterContinuationImpl<R> extends SuspensionPoint implements FilterContinuation<R> {
+    private abstract static class FilterContinuationImpl<R> implements FilterContinuation<R> {
+        /**
+         * This flag checks if the proceed method has been called, to request execute the next filter in chain.
+         */
+        final AtomicBoolean proceedRequested = new AtomicBoolean(false);
```

Comment:
these atomic bools just seem to guard the future .complete, you can instead check the return value of complete